docs/src/components/Demo/Demo.tsx

```diff
@@ -1,72 +1,119 @@
+      window.clearTimeout(newTimeout);
+      setCopyTimeout(0);
+    }, 2000);
+    window.clearTimeout(copyTimeout);
+    setCopyTimeout(newTimeout);
+    /* eslint-enable no-restricted-syntax */
+  }, [copyTimeout]);
+
+  const demo = useDemo(demoProps, {
+    copy: { onCopied },
+    defaultOpen,
+    export: exportOpts,
+    exportCodeSandbox,
+  });
+
+  const [fallbackToCodeSandbox, setFallbackToCodeSandbox] = React.useState(false);
+  React.useEffect(() => {
+    if (isSafari || isEdge) {
+      setFallbackToCodeSandbox(true);
+    }
+  }, []);
 
   return (
-    <BaseDemo.Root className={clsx('DemoRoot', className)} {...props}>
-      <DemoPlayground />
-      <Collapsible.Root open={open} onOpenChange={setOpen}>
+    <div className={clsx('DemoRoot', className)}>
+      {demo.allFilesSlugs.map(({ slug }) => (
+        <span key={slug} id={slug} className="scroll-mt-4" />
+      ))}
+      <DemoPlayground component={demo.component} variant={demo.selectedVariant} />
+      <Collapsible.Root open={demo.expanded} onOpenChange={demo.setExpanded}>
         <div role="figure" aria-label="Component demo code">
-          {(compact ? open : true) && (
+          {(compact ? demo.expanded : true) && (
             <div className="DemoToolbar">
-              <DemoFileSelector onTabChange={expandCollapsible} />
+              <DemoFileSelector
+                files={demo.files}
+                selectedFileName={demo.selectedFileName}
+                selectFileName={demo.selectFileName}
+                onTabChange={demo.expand}
+              />
 
               <div className="ml-auto flex items-center gap-4">
-                <DemoVariantSelector className="contents" onVariantChange={expandCollapsible} />
-                <CodeSandboxLink title="Base UI example" description="Base UI example" />
-                <BaseDemo.SourceCopy
-                  aria-label="Copy code"
-                  render={<GhostButton />}
-                  onCopied={() => {
-                    /* eslint-disable no-restricted-syntax */
-                    const newTimeout = window.setTimeout(() => {
-                      window.clearTimeout(newTimeout);
-                      setCopyTimeout(0);
-                    }, 2000);
-                    window.clearTimeout(copyTimeout);
-                    setCopyTimeout(newTimeout);
-                    /* eslint-enable no-restricted-syntax */
-                  }}
-                >
+                <DemoVariantSelector
+                  className="contents"
+                  onVariantChange={demo.expand}
+                  variants={demo.variants}
+                  selectedVariant={demo.selectedVariant}
+                  selectVariant={demo.selectVariant}
+                  availableTransforms={demo.availableTransforms}
+                  selectedTransform={demo.selectedTransform}
+                  selectTransform={demo.selectTransform}
+                />
+                {fallbackToCodeSandbox ? (
```

Comment:
Is there a problem with StackBlitz on Edge or Safari? 🤔 
Could we document the problem?

Reply:
Ok, gotcha. 👍 
Sad to see it. 🙈